commit/chainfee/outcome.go

```diff
@@ -0,0 +1,130 @@
+package chainfee
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+	"golang.org/x/exp/maps"
+
+	"github.com/smartcontractkit/chainlink-ccip/internal/libs/mathslib"
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugincommon"
+	"github.com/smartcontractkit/chainlink-ccip/internal/plugintypes"
+)
+
+func (p *processor) getConsensusObservation(
+	aos []plugincommon.AttributedObservation[Observation],
+) (ConsensusObservation, error) {
+	aggObs := aggregateObservations(aos)
+
+	fMin := mathslib.RepeatedF(func() int { return p.bigF }, maps.Keys(aggObs.FChain))
+	// consensus on the fChain map uses the role DON F value
+	// because all nodes can observe the home chain.
+	fChains := plugincommon.GetConsensusMap(p.lggr, "fChain", aggObs.FChain, fMin)
+
+	fDestChain, exists := fChains[p.destChain]
+	if !exists {
+		return ConsensusObservation{},
+			fmt.Errorf("no consensus value for fDestChain, destChain: %d", p.destChain)
+	}
+
+	timestamp := plugincommon.Median(aggObs.Timestamps, plugincommon.TimestampComparator)
+	chainFeeUpdatesConsensus := plugincommon.GetConsensusMapAggregator(
+		p.lggr,
+		"ChainFeePriceUpdates",
+		aggObs.ChainFeePriceUpdates,
+		mathslib.RepeatedF(
+			func() int { return mathslib.TwoFPlus1(fDestChain) },
+			maps.Keys(aggObs.ChainFeePriceUpdates),
+		),
+		plugincommon.TimestampedBigAggregator,
+	)
+
+	// Stop early if earliest updated timestamp is still fresh
+	earliestUpdateTime := plugincommon.EarliestTimestamp(maps.Values(chainFeeUpdatesConsensus), timestamp)
```

Comment:
I would expect `EarliestTimestamp` just returns the minimum from the provided slice.
I don't understand what `EarliestTimestamp` is doing with the second argument (`timestamp`).